Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp

```diff
@@ -162,50 +163,89 @@ void ABenchmarkGymGameModeBase::TickPlayersConnectedCheck(float DeltaSeconds)
 		return;
 	}
 
-	if (SecondsTillPlayerCheck > 0.0f)
+	// Only check players once
+	if (bHasDonePlayerCheck)
+	{
+		return;
+	}
+
+	const UNFRConstants* Constants = UNFRConstants::Get(GetWorld());
+	check(Constants);
+
+	// This test respects the initial delay timer in both native and GDK
+	if (Constants->PlayerCheckValid())
 	{
-		SecondsTillPlayerCheck -= DeltaSeconds;
-		if (SecondsTillPlayerCheck <= 0.0f)
+		bHasDonePlayerCheck = true;
+		if (ActivePlayers != ExpectedPlayers)
 		{
-			if (ActivePlayers != ExpectedPlayers)
-			{
-				// This log is used by the NFR pipeline to indicate if a client failed to connect
-				UE_LOG(LogBenchmarkGymGameModeBase, Error, TEXT("A client connection was dropped. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
-			}
-			else
-			{
-				// Useful for NFR log inspection
-				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("All clients successfully connected. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
-			}
+			// This log is used by the NFR pipeline to indicate if a client failed to connect
+			UE_LOG(LogBenchmarkGymGameModeBase, Error, TEXT("A client connection was dropped. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
+		}
+		else
+		{
+			// Useful for NFR log inspection
+			UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("All clients successfully connected. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
 		}
 	}
 }
 
+void ABenchmarkGymGameModeBase::FailServerFPSCheck(const float FPS)
+{
+	bHasFpsFailed = true;
+	NFR_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("FPS check failed. FPS: %.8f"), FPS);
+}
+
 void ABenchmarkGymGameModeBase::TickServerFPSCheck(float DeltaSeconds)
 {
-	const UNFRConstants* Constants = UNFRConstants::Get(GetWorld());
+	// Return if we have already failed
+	if (bHasFpsFailed)
+	{
+		return;
+	}
+
+	const UWorld* World = GetWorld();
+	const UGDKTestGymsGameInstance* GameInstance = Cast<UGDKTestGymsGameInstance>(World->GetGameInstance());
+	if (GameInstance == nullptr)
+	{
+		return;
+	}
+
+	const UNFRConstants* Constants = UNFRConstants::Get(World);
 	check(Constants);
-	if (Constants->SamplesForFPSValid() && !bHasFpsFailed && GetWorld() != nullptr)
+
+	const float FPS = GameInstance->GetAveragedFPS();
+	if (FPS < Constants->GetMinServerFPS())
 	{
-		if (const UGDKTestGymsGameInstance* GameInstance = Cast<UGDKTestGymsGameInstance>(GetWorld()->GetGameInstance()))
+		if (IsUsingSpatialNetworking)
```

Comment:
Could be worth adding a comment here about why we immediately fail for spatial

Reply:
Also I don't think this behavior matches now we use the boolean fps check

Reply:
Think wrapping in `SamplesForServerFPSValid` should fix it, we want to upload 1 until we start actually checking

Reply:
bHasFpsFailed is false by default so GetFPSValid will return 1 by default. I think this matches?

Reply:
Oops yeah, might still be slightly incorrect though - we can fail from the start of the deployment which doesn't match native. It's slightly different for this because we use the metric as a signal instead of an actual value so the 2 minute window is on the signal not the FPS check (for which the signal only goes from 1->0 which means the deployment has failed)